otherlibs/win32unix/unixsupport.c

```diff
@@ -328,3 +328,140 @@ int unix_cloexec_p(value cloexec)
   else
     return unix_cloexec_default;
 }
+
+BOOL unix_drop_privilege(DWORD dwAttributes,
+                         LPCWSTR lpName,
+                         PHANDLE hToken_out,
+                         PTOKEN_PRIVILEGES restore,
+                         PDWORD dwFlags)
+{
+  BOOL result = TRUE;
+  if (((dwAttributes & FILE_ATTRIBUTE_DIRECTORY) != 0)) {
+    LUID privilege_luid;
+    DWORD dwReturnLength;
+    HANDLE hToken;
+
+    /* FILE_FLAG_BACKUP_SEMANTICS is required to get a handle to a directory */
+    *dwFlags |= FILE_FLAG_BACKUP_SEMANTICS;
+
+    /* However, we don't want to access a directory which we can only see
+       through SeRestorePrivilege, so ensure it's disabled when CreateFile is
+       called and restore it if necessary afterwards.
+
+       As always, this isn't easy...
+     */
+
+    /* Step 1: get the effective privileges for this thread */
+    LookupPrivilegeValue(NULL, lpName, &privilege_luid);
+    /* -6 is the pseudo-handle for GetCurrentThreadEffectiveToken, which is not
+       present in the mingw-w64 headers */
+    if (!GetTokenInformation((HANDLE)(LONG_PTR)-6,
+                             TokenPrivileges,
+                             NULL, 0, &dwReturnLength)
+        && GetLastError() == ERROR_INSUFFICIENT_BUFFER) {
+      TOKEN_PRIVILEGES* privs;
+
+      if (!(privs = (TOKEN_PRIVILEGES*)malloc(dwReturnLength)))
+        caml_raise_out_of_memory();
+
+      if (GetTokenInformation((HANDLE)(LONG_PTR)-6,
+                              TokenPrivileges, privs,
+                              dwReturnLength, &dwReturnLength)) {
+        int i = 0;
+        LUID_AND_ATTRIBUTES* privilege = privs->Privileges;
+
+        /* Search for SeRestorePrivilege */
+        while (i < privs->PrivilegeCount
+               && (privilege->Luid.HighPart != privilege_luid.HighPart
+                   || privilege->Luid.LowPart != privilege_luid.LowPart)) {
+          i++;
+          privilege++;
+        }
+
+        if (i < privs->PrivilegeCount) {
```

Comment:
I think that if we have a token and the token does not mention SeRestorePrivilege, then this condition is false, `restore->PrivilegeCount` will never be initialised, the function returns `TRUE`, and a later `unix_restore_privilege` reads the uninitialised `restore->PrivilegeCount`. I think?